File: backend/src/main/java/com/culinaryheaven/domain/comment/service/CommentService.java
package com.culinaryheaven.domain.comment.service;

import com.culinaryheaven.domain.comment.domain.Comment;
import com.culinaryheaven.domain.comment.dto.request.CommentCreateRequest;
import com.culinaryheaven.domain.comment.dto.response.CommentResponse;
import com.culinaryheaven.domain.comment.dto.response.CommentsResponse;
import com.culinaryheaven.domain.comment.repository.CommentRepository;
import com.culinaryheaven.domain.recipe.domain.Recipe;
import com.culinaryheaven.domain.recipe.repository.RecipeRepository;
import com.culinaryheaven.domain.user.domain.User;
import com.culinaryheaven.domain.user.repository.UserRepository;
import com.culinaryheaven.global.exception.CustomException;
import com.culinaryheaven.global.exception.ErrorCode;
import com.culinaryheaven.global.util.SecurityUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CommentService {

    private final UserRepository userRepository;
    private final RecipeRepository recipeRepository;
    private final CommentRepository commentRepository;
    private final SecurityUtil securityUtil;

    @Transactional
    public CommentResponse createComment(CommentCreateRequest request) {

        User user = userRepository.findByOauthId(
                securityUtil.getUserOAuth2Id()
        ).orElseThrow(() -> new CustomException(ErrorCode.AUTHORIZATION_FAILED));

        Recipe recipe = recipeRepository.findById(request.recipeId()).orElseThrow(
                () -> new CustomException(ErrorCode.RECIPE_NOT_FOUND));

        Comment comment = request.toEntity(user, recipe);
        Comment savedComment = commentRepository.save(comment);

        return CommentResponse.of(savedComment);

    }

    public CommentsResponse getCommentsByRecipeId(Long recipeId) {

        List<Comment> comments = commentRepository.findByRecipeId(
                Sort.by(Sort.Direction.DESC, "createdAt"), recipeId
        );

        return CommentsResponse.of(comments);

    }

}

File: android/data/src/main/java/com/jae464/data/remote/api/CommentService.kt
package com.jae464.data.remote.api

import com.jae464.data.remote.model.request.comment.CommentCreateRequest
import com.jae464.data.remote.model.request.comment.CommentUpdateRequest
import com.jae464.data.remote.model.response.comment.CommentResponse
import com.jae464.data.remote.model.response.comment.CommentsResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query

interface CommentService {

    @POST("/comments")
    suspend fun createComment(
        @Body request: CommentCreateRequest,
    ): Response<CommentResponse>

    @GET("/comments")
    suspend fun getCommentsByRecipeId(
        @Query("recipeId") recipeId: Long,
    ): Response<CommentsResponse>

    @PATCH("/comments/{commentId}")
    suspend fun updateComment(
        @Path("commentId") commentId: Long,
        @Body request: CommentUpdateRequest,
    ): Response<CommentResponse>

    @DELETE("/comments/{commentId}")
    suspend fun deleteComment(
        @Path("commentId") commentId: Long,
    ): Response<Unit>

}